src/Microsoft.FeatureManagement/ContextualFeatureFilterEvaluator.cs

```diff
@@ -0,0 +1,99 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+//
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace Microsoft.FeatureManagement
+{
+    /// <summary>
+    /// Provides a performance efficient method of evaluating IContextualFeatureFilter&lt;T&gt; without knowing what the generic type parameter is.
+    /// </summary>
+    class ContextualFeatureFilterEvaluator : IContextualFeatureFilter<object>
+    {
+        private IFeatureFilterMetadata _filter;
+        private Func<object, FeatureFilterEvaluationContext, object, Task<bool>> _evaluateFunc;
+
+        public ContextualFeatureFilterEvaluator(IFeatureFilterMetadata filter, Type appContextType)
+        {
+            if (filter == null)
+            {
+                throw new ArgumentNullException(nameof(filter));
+            }
+
+            if (appContextType == null)
+            {
+                throw new ArgumentNullException(nameof(appContextType));
+            }
+
+            Type targetInterface = GetContextualFilterInterface(filter, appContextType);
+
+            //
+            // Extract IContextualFeatureFilter<T>.EvaluateAsync method.
+            if (targetInterface != null)
+            {
+                MethodInfo evaluateMethod = targetInterface.GetMethod(nameof(IContextualFeatureFilter<object>.EvaluateAsync), BindingFlags.Public | BindingFlags.Instance);
+
+                _evaluateFunc = TypeAgnosticEvaluate(filter.GetType(), evaluateMethod);
+            }
+
+            _filter = filter;
+        }
+
+        public static Type GetContextualFilterInterface(IFeatureFilterMetadata filter, Type appContextType)
+        {
+            IEnumerable<Type> contextualFilterInterfaces = filter.GetType().GetInterfaces().Where(i => i.IsGenericType && i.GetGenericTypeDefinition().IsAssignableFrom(typeof(IContextualFeatureFilter<>)));
+
+            Type targetInterface = null;
+
+            if (contextualFilterInterfaces != null)
+            {
+                targetInterface = contextualFilterInterfaces.FirstOrDefault(i => i.GetGenericArguments()[0].IsAssignableFrom(appContextType));
```

Comment:
>FirstOrDefault [](start = 61, length = 14)

if user's filter implement two IContextualFeatureFilter<A> and IContextualFeatureFilter<B> and class A : B {}. In this case, we might not just get the first but the best matched.

Reply:
That scenario is quite confusing. I'm not sure it could be clear what to expect there. I think it would be better not to support that.

---
In reply to: [328395534](https://github.com/microsoft/FeatureManagement-Dotnet/pull/14#discussion_r328395534) [](ancestors = 328395534)